File: src/mechanism/qf.py
from __future__ import annotations

import math


class QFEngine:
    """Standard Quadratic Funding calculation with optional cap."""

    def __init__(self, cap_per_project: float = 0.25):
        self.cap_per_project = cap_per_project

    def calculate(self, contributions: dict[str, list[float]], matching_pool: float) -> dict[str, float]:
        if not contributions:
            return {}

        qf_scores: dict[str, float] = {}
        for project, contribs in contributions.items():
            cleaned = [max(0.0, float(c)) for c in contribs if c is not None]
            sqrt_sum = sum(math.sqrt(c) for c in cleaned)
            qf_scores[project] = float(sqrt_sum ** 2)

        total = sum(qf_scores.values())
        if total <= 0:
            equal = matching_pool / len(contributions)
            return {p: equal for p in contributions}

        alloc = {p: (score / total) * matching_pool for p, score in qf_scores.items()}
        return self._apply_cap(alloc, matching_pool)

    def _apply_cap(self, allocations: dict[str, float], matching_pool: float) -> dict[str, float]:
        cap = matching_pool * self.cap_per_project
        if not allocations:
            return {}

        capped = dict(allocations)
        for _ in range(10):
            over = {p: v for p, v in capped.items() if v > cap}
            if not over:
                return capped
            excess = sum(v - cap for v in over.values())
            for p in over:
                capped[p] = cap
            under = [p for p, v in capped.items() if v < cap]
            if not under or excess <= 0:
                break
            under_total = sum(capped[p] for p in under)
            if under_total <= 0:
                share = excess / len(under)
                for p in under:
                    capped[p] += share
            else:
                for p in under:
                    capped[p] += excess * (capped[p] / under_total)
        return capped

File: packages/backend/src/lib/qf.ts
/**
 * Quadratic Funding (QF) Engine
 * Ported from /shared/simogrants/src/mechanism/qf.py
 */

export class QFEngine {
  private capPerProject: number;

  constructor(capPerProject: number = 0.25) {
    this.capPerProject = capPerProject;
  }

  /**
   * Calculate QF allocations from contributions.
   * 
   * For each project: sqrt_sum = sum(sqrt(contribution)) for each contributor
   * QF score = sqrt_sum^2
   * allocation = (score / total_scores) * matching_pool
   */
  calculate(
    contributions: Map<string, number[]>,
    matchingPool: number
  ): Map<string, number> {
    if (contributions.size === 0) return new Map();

    const qfScores = new Map<string, number>();

    for (const [project, contribs] of contributions) {
      const cleaned = contribs
        .filter((c) => c != null)
        .map((c) => Math.max(0.0, Number(c)));
      const sqrtSum = cleaned.reduce((sum, c) => sum + Math.sqrt(c), 0);
      qfScores.set(project, sqrtSum * sqrtSum);
    }

    let total = 0;
    for (const score of qfScores.values()) total += score;

    if (total <= 0) {
      const equal = matchingPool / contributions.size;
      return new Map(
        Array.from(contributions.keys()).map((p) => [p, equal])
      );
    }

    const alloc = new Map<string, number>();
    for (const [project, score] of qfScores) {
      alloc.set(project, (score / total) * matchingPool);
    }

    return this.applyCap(alloc, matchingPool);
  }

  private applyCap(
    allocations: Map<string, number>,
    matchingPool: number
  ): Map<string, number> {
    const cap = matchingPool * this.capPerProject;
    if (allocations.size === 0) return new Map();

    const capped = new Map(allocations);

    for (let i = 0; i < 10; i++) {
      const over: string[] = [];
      for (const [p, v] of capped) {
        if (v > cap) over.push(p);
      }
      if (over.length === 0) return capped;

      let excess = 0;
      for (const p of over) {
        excess += capped.get(p)! - cap;
        capped.set(p, cap);
      }

      const under: string[] = [];
      for (const [p, v] of capped) {
        if (v < cap) under.push(p);
      }

      if (under.length === 0 || excess <= 0) break;

      const underTotal = under.reduce((sum, p) => sum + capped.get(p)!, 0);
      if (underTotal <= 0) {
        const share = excess / under.length;
        for (const p of under) capped.set(p, capped.get(p)! + share);
      } else {
        for (const p of under) {
          capped.set(p, capped.get(p)! + excess * (capped.get(p)! / underTotal));
        }
      }
    }

    return capped;
  }
}
